### ISSUE TITLE
DDS late initialization errors

### ISSUE THREAD

**Issue by zanderso** [flutter/flutter#86361]
From Flutter CLI crash logging on dev channel dev/2.4.0-0.0.pre, we're seeing these crashes from dds:
```
Thread 0 main threadLateError: LateInitializationError: Field '_server@412427471' has not been initialized.
at DartDevelopmentServiceImpl._server(dds_impl.dart)
at DartDevelopmentServiceImpl.shutdown(dds_impl.dart:173)
at DartDevelopmentServiceImpl.startService.<anonymous closure>(dds_impl.dart:86)
at _rootRunBinary(zone.dart:1452)
at _CustomZone.runBinary(zone.dart:1342)
at _FutureListener.handleError(future_impl.dart:175)
at Future._propagateToListeners.handleError(future_impl.dart:779)
at Future._propagateToListeners(future_impl.dart:800)
at Future._completeError(future_impl.dart:610)
at Future.wait.<anonymous closure>(future.dart:413)
at _rootRunUnary(zone.dart:1436)
at _CustomZone.runUnary(zone.dart:1335)
at <asynchronous gap>(async)
at Future.wait.<anonymous closure>(future.dart)
at <asynchronous gap>(async)
```

and similarly
```
Thread 0 main threadLateError: LateInitializationError: Field '_server@412427471' has not been initialized.
at DartDevelopmentServiceImpl._server(dds_impl.dart)
at DartDevelopmentServiceImpl.shutdown(dds_impl.dart:173)
at DartDevelopmentServiceImpl.startService.<anonymous closure>(dds_impl.dart:79)
at _rootRunUnary(zone.dart:1436)
at _CustomZone.runUnary(zone.dart:1335)
at <asynchronous gap>(async)
at DartDevelopmentServiceImpl.startService.<anonymous closure>(dds_impl.dart)
at <asynchronous gap>(async)
```

@bkonyi

**Comment by zanderso** [comment]
Bumping up priority, and marking for cherry-pick into 2.4 because this will noticeably elevate the tool crash rate.

**Comment by bkonyi** [comment]
Should be an easy fix. Will try to get a fix published today.

**Comment by bkonyi** [comment]
Fixed in `2.0.2`.

**Comment by zanderso** [comment]
@christopherfujino When you get a chance, could you try updating flutter_tool to this version?

**Comment by christopherfujino** [comment]
> @christopherfujino When you get a chance, could you try updating flutter_tool to this version?

will do

**Comment by zanderso** [comment]
This is still happening in dev/2.5.0-5.0.pre
```
Thread 0 main threadLateError: LateInitializationError: Field '_server@418427471' has not been initialized.
at DartDevelopmentServiceImpl._server(dds_impl.dart)
at DartDevelopmentServiceImpl.shutdown(dds_impl.dart:182)
at DebugService.close(debug_service.dart:145)
at AppDebugServices.close(app_debug_services.dart:34)
at DevHandler.loadAppServices.<anonymous closure>(dev_handler.dart:229)
at DevHandler.loadAppServices.<anonymous closure>(dev_handler.dart:228)
at _rootRun(zone.dart:1420)
at _CustomZone.run(zone.dart:1328)
at _FutureListener.handleWhenComplete(future_impl.dart:203)
at Future._propagateToListeners.handleWhenCompleteCallback(future_impl.dart:737)
at Future._propagateToListeners(future_impl.dart:793)
at Future._complete(future_impl.dart:593)
at _cancelAndValue(stream_pipe.dart:61)
at Stream.first.<anonymous closure>(stream.dart:1232)
at _rootRunUnary(zone.dart:1436)
at _CustomZone.runUnary(zone.dart:1335)
at _CustomZone.runUnaryGuarded(zone.dart:1244)
at _BufferingStreamSubscription._sendData(stream_impl.dart:341)
at _DelayedData.perform(stream_impl.dart:591)
at _StreamImplEvents.handleNext(stream_impl.dart:706)
at _PendingEvents.schedule.<anonymous closure>(stream_impl.dart:663)
at _rootRun(zone.dart:1420)
at _CustomZone.run(zone.dart:1328)
at _CustomZone.runGuarded(zone.dart:1236)
at _CustomZone.bindCallbackGuarded.<anonymous closure>(zone.dart:1276)
at _rootRun(zone.dart:1428)
at _CustomZone.run(zone.dart:1328)
at _CustomZone.runGuarded(zone.dart:1236)
at _CustomZone.bindCallbackGuarded.<anonymous closure>(zone.dart:1276)
at _microtaskLoop(schedule_microtask.dart:40)
at _startMicrotaskLoop(schedule_microtask.dart:49)
at _runPendingImmediateCallback(isolate_patch.dart:120)
at _RawReceivePortImpl._handleMessage(isolate_patch.dart:185)
```

**Comment by zanderso** [comment]
flutter_tools has dds at 2.0.2: https://github.com/flutter/flutter/blob/master/packages/flutter_tools/pubspec.yaml#L13

**Comment by bkonyi** [comment]
Very strange that we're seeing a `LateInitializationError` for `_server` not coming from a call to `shutdown()` from within DDS since we should only ever return a `DartDevelopmentService` instance after initialization is completely finished... I wonder if there's a subtle issue related to the order of execution in the event queue after an error occurs?

@zra is this something we're seeing frequently or is this a one-off?

**Comment by christopherfujino** [comment]
> @zanderso is this something we're seeing frequently or is this a one-off?

We don't have enough usage to get a good feel for that, but it's the top crasher.

**Comment by zanderso** [comment]
@bkonyi Could you publish a new version of `package:dds` please?

**Comment by bkonyi** [comment]
`2.1.1` has been released.

**Comment by christopherfujino** [comment]
so it looks like the tool cannot upgrade `dds` from `2.0.2 -> 2.1.1` because the latter depends on `vm_service 7.2.0`, while the newest `dwds` depends on `vm_service 7.1.1`.

In order for the tool to apply this change, either `dwds` would need to bump their dependency of `vm_service` or `dds` would need to put out a special hotfix version that does not include the vm_service bump. I would prefer the latter as it would make cherrypicking to the beta branch easier.

**Comment by christopherfujino** [comment]
cc @annagrin

**Comment by annagrin** [comment]
@christopherfujino I can  bump the version and release the new dwds, just in option 2 is not doable 
/cc @bkonyi

**Comment by annagrin** [comment]
Released dwds 11.2.2

**Comment by bkonyi** [comment]
Published a `2.0.3` version without the `vm_service` version bump.

**Comment by christopherfujino** [comment]
> Published a `2.0.3` version without the `vm_service` version bump.

awesome, thanks Ben!

**Comment by bkonyi** [comment]
Should be resolved now that #88728 has landed.
